packages/web3torrent/src/contexts/omni-provider.tsx

```diff
@@ -0,0 +1,21 @@
+import {BudgetContext, useBudgetContext} from './budget-context';
+import {InitializationContext, useInitializationContext} from './initialization-context';
+import React from 'react';
+import {ChannelContext, useChannelContext} from './channel-context';
+import {TorrentContext, useTorrentContext} from './torrent-context';
+
+export const OmniProvider: React.FC = ({children}) => {
+  const initializationContext = useInitializationContext();
+  // TODO: Probably a more elegant way of doing this
+  return (
+    <InitializationContext.Provider value={initializationContext}>
+      <TorrentContext.Provider value={useTorrentContext({initializationContext})}>
+        <ChannelContext.Provider value={useChannelContext({initializationContext})}>
+          <BudgetContext.Provider value={useBudgetContext({initializationContext})}>
```

Comment:
Is it true that the values of these providers will be updated only when the `OmniProvider` is re-rendered? If so, is the main purpose of finegrain providers to wire finegrain data to the component that needs the data?

My understanding is that when `OmniProvider` component is update, the entire React tree needs to be re-rendered. So I am not sure that wiring finegrain data to components is a performance gain. Though I am not 100% sure about that.

Reply:
> Is it true that the `value` properties of these providers will be updated only when the `OmniProvider` is re-rendered? If so, is the main purpose of finegrain providers to wire finegrain data to the component that needs the data?

I think it's kind of the opposite way around. Whenever the `value` changes it forces a rerender. My understanding is that `react` is smart enough to only render components that are affected by a context change.

From [react docs](https://reactjs.org/docs/context.html#contextprovider):
 >All consumers that are descendants of a Provider will re-render whenever the Provider’s value prop changes"

A consumer is a component calling the `useContext` hook or the old style
```
<MyContext.Consumer>
  {value => /* render something based on the context value */}
</MyContext.Consumer>
```

In our case a consumer is something like  `useContext(BudgetContext)`. `react` is smart enough( if my understanding is correct) to rerender whatever components call `useContext(BudgetContext)`

So by having narrower contexts you would only cause rerenders in specific components.

All that being said, our react app is pretty simple (from a UI perspective) so I don't think the performance aspect really matters much. (And I haven't really bothered to look at exactly when things are being rerendered)

My preference for narrower contexts is that it seems more natural in the components to access a narrow piece of data, instead of having to navigate some complex context object.

Reply:
I agree with all of your points, and your understand of React matches mine. I think it is true that `BudgetContext.Provider` `value` property is only updated when `OmniProvider` is updated.

Also, I think [this caveat](https://reactjs.org/docs/context.html#caveats) is something to be aware of. For us, I think this means that the `value` property will be updated every time `OmniProvider` is re-rendered.

Reply:
I gave this some more thought, and I think the following will happen:
1. Something triggers a state update for `OmniProvider` component. This can be one of the children components calling a `setState` function provided by a contexts.
1. The OmniProvider is re-rendered. During re-rendering, every `value` property for every context is recomputed.
1. All context consumers are re-rendered due to `value` updates.

Not a big deal given that the application is fairly small. But I think the implication is that every component that is a context consumer is re-rendered every time `OmniProvider` component is re-rendered (which happens any time finegrain state-setter is called).

Reply:
Good point! I hadn't really considered that `OmniProvider` owns all the various states of the custom hooks it's using, which forces a rerender when changed.

I think we could avoid this by not custom hooks to construct a context . I'm going to take a second look at if we even need to write the context as custom hooks.

Reply:
> I think we could avoid this by not custom hooks to construct a context . I'm going to take a second look at if we even need to write the context as custom hooks.

So I've spent some time looking at just doing this with classes for the context and it seems more awkward than using these custom hooks.

Reply:
I have tossing ideas around my head and trying to distill the essence of this code. So I think it is:
1. We would like to define state at the top level of the component tree instead of leaves.
1. We would like the ability to pass this state and corresponding state setters down the leaves without painful prop chaining.
1. We would like this state to be more finegrain as opposed to a large global state object. 
  1. This is done to limit the scope of knowledge for a component. 
  1. And the hope with this was to optimize re-rending.

Generic contexts give us #1 and #2. Seems like 3b will be tricky to get. Then I guess it all comes down to how strongly we feel about 3a.